mobile/src/main/java/ch/epfl/sdp/mobile/ui/game/Confetti.kt

```diff
@@ -0,0 +1,240 @@
+
+package ch.epfl.sdp.mobile.ui.game
+
+import androidx.compose.animation.core.*
+import androidx.compose.runtime.*
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.draw.drawWithContent
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.geometry.Size
+import androidx.compose.ui.geometry.center
+import androidx.compose.ui.graphics.Color
+import androidx.compose.ui.graphics.drawscope.ContentDrawScope
+import androidx.compose.ui.graphics.drawscope.withTransform
+import androidx.compose.ui.unit.DpOffset
+import androidx.compose.ui.unit.DpOffset.Companion.Zero
+import androidx.compose.ui.unit.dp
+import kotlin.math.PI
+import kotlin.math.cos
+import kotlin.math.sin
+import kotlin.random.Random
+import kotlinx.coroutines.NonCancellable
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
+
+/** Creates a [ConfettiState] and remembers it in the composition. */
+@Composable
+fun rememberConfettiState(): ConfettiState {
+  return remember { ConfettiState() }
+}
+
+/**
+ * A class describing the state of the confetti which are currently displayed on the screen.
+ * Spawning confetti is done with some suspending animation functions.
+ */
+@Stable
+class ConfettiState {
+
+  /** Ensures mutually exclusive access to [confetti]. */
+  private val mutex = Mutex()
+
+  /** The [Confetti] which should be rendered. */
+  private val confetti = mutableStateListOf<Confetti>()
+
+  /** Returns true if the animation is currently running. */
+  val isRunning: Boolean
+    get() = confetti.isNotEmpty()
+
+  /**
+   * Spawns a single confetti, and runs it in a suspending fashion.
+   *
+   * @param angle the angle which the spawn cone targets.
+   * @param spread the width of the spawn cone.
+   * @param colors the possible colors for the confetti.
+   */
+  private suspend fun spawnOne(angle: Float, spread: Float, colors: List<Color>) {
+    require(angle >= 0f && angle < 360f) { "Angle must be a valid angle." }
+    require(spread > 0f && spread <= 360f) { "Spread must be a valid angle." }
+    require(colors.isNotEmpty()) { "Must provide at least one confetti color." }
+
+    val firstAngle = angle.toDouble() - spread / 2f
+    val secondAngle = angle.toDouble() + spread / 2f
+    val fromAngle = minOf(firstAngle, secondAngle)
+    val untilAngle = maxOf(firstAngle, secondAngle)
+
+    val confettiAngle = Random.nextDouble(fromAngle, untilAngle).toFloat()
+    val confettiDistance = Random.nextDouble(from = 12.0, until = 36.0)
+
+    val x = (cos(2 * PI * confettiAngle / 360f) * confettiDistance).dp
+    val y = (-sin(2 * PI * confettiAngle / 360f) * confettiDistance).dp
+
+    val state =
+        Confetti(
+            color = colors.random(),
+            initialSpeed = DpOffset(x * 30, y * 30),
+            targetSlide = DpOffset(0.dp, 180.dp),
```

Comment:
Here as well.

Reply:
Extracted to default params in 7e5b878c4117c72514b5d7c617ff6c7fa511c0b1